res/controllers/M-Audio-Xponent-Advanced-scripts.js

```diff
@@ -0,0 +1,842 @@
+        engine.trigger("[EqualizerRack1_" + deck.group + "_Effect1]", "button_parameter3");
+
+        engine.trigger(deck.group, "keylock");
+        engine.trigger(deck.group, "hotcue_1_enabled");
+        engine.trigger(deck.group, "hotcue_2_enabled");
+        engine.trigger(deck.group, "hotcue_3_enabled");
+        engine.trigger(deck.group, "hotcue_4_enabled");
+        engine.trigger(deck.group, "hotcue_5_enabled");
+
+        engine.trigger(deck.group, "loop_in");
+        engine.trigger(deck.group, "loop_out");
+
+        midi.sendShortMsg(deck.on, MaudioXponent.buttons.scratch, deck.scratchEnabled);
+        MaudioXponent.onPlayPositionChange(deck.playPosition, deck.group);
+    }
+};
+
+MaudioXponent.convert = function(value) {
+    return (value * 127).toFixed(0);
+};
+
+MaudioXponent.getDeck = function(group) {
+    return MaudioXponent.decks[parseInt(group.substring(8)) - 1];
+};
+
+MaudioXponent.pauseScript = function(ms) {
+    startDate = new Date();
+    currentDate = null;
+    while(currentDate-startDate < ms) currentDate = new Date();
+};
+
+MaudioXponent.vuMeter = function(mode, bank, side, value) {
+    if (mode === MaudioXponent.config.vuMeterMode) { // Accept messages for the active mode
+        if ((mode === 0) || (bank === MaudioXponent.state.bank)) { // Accept messages for the active bank
+            value = MaudioXponent.convert(value);
+
+            if (side === 0) {
+                midi.sendShortMsg(MaudioXponent.leftDeck.vuMeterStatusByte, MaudioXponent.leftDeck.vuMeterSecondByte, value);
+            } else {
+                midi.sendShortMsg(MaudioXponent.rightDeck.vuMeterStatusByte, MaudioXponent.rightDeck.vuMeterSecondByte, value);
+            }
+        }
+    }
+}
+
+MaudioXponent.bankSwitch = function(channel, control, value, status, group) {
+    //script.midiDebug(channel, control, value, status, group);
+    MaudioXponent.state.bank = status;
+
+    if (status === MaudioXponent.bankA) {
+        MaudioXponent.leftDeck = MaudioXponent.decks[0];
+        MaudioXponent.rightDeck = MaudioXponent.decks[1];
+    } else if (status === MaudioXponent.bankB) {
+        MaudioXponent.leftDeck = MaudioXponent.decks[2];
+        MaudioXponent.rightDeck = MaudioXponent.decks[3];
+    }
+
+    MaudioXponent.syncLights();
+};
+
+MaudioXponent.beatsync = function(channel, control, value, status, group) {
+    //script.midiDebug(channel, control, value, status, group);
+    var deck = MaudioXponent.getDeck(group);
+    var activate = (status == deck.on);
+
+    if (activate) {
+        if (MaudioXponent.leftDeck.shift && MaudioXponent.rightDeck.shift) {
+            // Double-shift = cycle modes
+            var mode = MaudioXponent.config.syncFlashMode + 1;
+            if(mode===3){
+                mode = 0;
+            }
+            MaudioXponent.config.syncFlashMode = mode;
+        } else {
+            if (!deck.shift) {
+                engine.setValue(deck.group, "beatsync", 0x04);
```

Comment:
I can see that. Unless a long-press is handled by the engine, I'm not sure I'd want to do that in the script. I don't want javascript responsible for polling, timing, any of that. I could change the shift-sync behavior, and reassign the tap behavior, but it seems that it doesn't work anymore anyway, even from the UI. If a track has existing bpm information, it won't listen to the tap, so I 'll just repurpose the button.